File: mobile/modules/crust/android/src/main/java/com/mentra/crust/CrustModule.kt
package com.mentra.crust

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL

class CrustModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Crust")

    Constant("PI") {
      Math.PI
    }

    Events(
      "onChange",
      "onNavManeuver",
      "onNavRerouting",
      "onNavArrived",
      "onNavError",
      "onNavLocation",
      "onNavRoute",
      "onHeading",
    )

    Function("hello") {
      "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { value: String ->
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    View(CrustView::class) {
      Prop("url") { view: CrustView, url: URL ->
        view.webView.loadUrl(url.toString())
      }
      Events("onLoad")
    }

    // MARK: - Image Processing Commands

    AsyncFunction("processGalleryImage") { inputPath: String, outputPath: String, options: Map<String, Any?> ->
      try {
        val inputFile = java.io.File(inputPath)
        if (!inputFile.exists()) {
          throw IllegalArgumentException("Input file does not exist: $inputPath")
        }

        val lensCorrection = options["lensCorrection"] as? Boolean ?: true
        val colorCorrection = options["colorCorrection"] as? Boolean ?: true

        val processingTimeMs = com.mentra.crust.utils.ImageProcessor.process(
          inputPath, outputPath, lensCorrection, colorCorrection, 95
        )

        if (processingTimeMs >= 0) {
          mapOf(
            "success" to true,
            "outputPath" to outputPath,
            "processingTimeMs" to processingTimeMs
          )
        } else {
          mapOf("success" to false, "error" to "Processing failed")
        }
      } catch (e: Exception) {
        android.util.Log.e("CrustModule", "processGalleryImage error: ${e.message}", e)
        mapOf("success" to false, "error" to (e.message ?: "Unknown error"))
      }
    }

    // MARK: - HDR Merge Commands

    AsyncFunction("mergeHdrBrackets") { underPath: String, normalPath: String, overPath: String, outputPath: String ->
      try {
        val processingTimeMs = com.mentra.crust.utils.ImageProcessor.mergeHdr(
          underPath, normalPath, overPath, outputPath, 95
        )
        if (processingTimeMs >= 0) {
          mapOf(
            "success" to true,
            "outputPath" to outputPath,
            "processingTimeMs" to processingTimeMs
          )
        } else {
          mapOf("success" to false, "error" to "HDR merge failed")
        }
      } catch (e: Exception) {
        android.util.Log.e("CrustModule", "mergeHdrBrackets error: ${e.message}", e)
        mapOf("success" to false, "error" to (e.message ?: "Unknown error"))
      }
    }

    // MARK: - Video Stabilization Commands

    AsyncFunction("stabilizeVideo") { inputPath: String, imuPath: String, outputPath: String ->
      try {
        val inputFile = java.io.File(inputPath)
        val imuFile = java.io.File(imuPath)
        if (!inputFile.exists()) {
          throw IllegalArgumentException("Input video does not exist: $inputPath")
        }
        if (!imuFile.exists()) {
          throw IllegalArgumentException("IMU sidecar does not exist: $imuPath")
        }

        val processingTimeMs = com.mentra.crust.utils.VideoStabilizer.stabilize(
          inputPath, imuPath, outputPath
        )

        if (processingTimeMs >= 0) {
          mapOf(
            "success" to true,
            "outputPath" to outputPath,
            "processingTimeMs" to processingTimeMs
          )
        } else {
          mapOf("success" to false, "error" to "Stabilization failed")
        }
      } catch (e: Exception) {
        android.util.Log.e("CrustModule", "stabilizeVideo error: ${e.message}", e)
        mapOf("success" to false, "error" to (e.message ?: "Unknown error"))
      }
    }

    // MARK: - Media Library Commands

    AsyncFunction("saveToGalleryWithDate") { filePath: String, captureTimeMillis: Long? ->
      val context =
        appContext.reactContext
          ?: appContext.currentActivity
            ?: throw IllegalStateException("No context available")

      try {
        val file = java.io.File(filePath)
        if (!file.exists()) {
          throw IllegalArgumentException("File does not exist: $filePath")
        }

        val mimeType =
          when (file.extension.lowercase()) {
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "mp4" -> "video/mp4"
            "mov" -> "video/quicktime"
            else -> "application/octet-stream"
          }

        val isVideo = mimeType.startsWith("video/")
        val collection =
          if (isVideo) {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
              android.provider.MediaStore.Video.Media.getContentUri(
                android.provider.MediaStore.VOLUME_EXTERNAL_PRIMARY
              )
            } else {
              android.provider.MediaStore.Video.Media.EXTERNAL_CONTENT_URI
            }
          } else {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
              android.provider.MediaStore.Images.Media.getContentUri(
                android.provider.MediaStore.VOLUME_EXTERNAL_PRIMARY
              )
            } else {
              android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI
            }
          }

        val values =
          android.content.ContentValues().apply {
            put(android.provider.MediaStore.MediaColumns.DISPLAY_NAME, file.name)
            put(android.provider.MediaStore.MediaColumns.MIME_TYPE, mimeType)
            put(android.provider.MediaStore.MediaColumns.SIZE, file.length())

            if (captureTimeMillis != null) {
              if (isVideo) {
                put(android.provider.MediaStore.Video.Media.DATE_TAKEN, captureTimeMillis)
              } else {
                put(android.provider.MediaStore.Images.Media.DATE_TAKEN, captureTimeMillis)
              }
              android.util.Log.d(
                "CrustModule",
                "Setting DATE_TAKEN to: $captureTimeMillis (${java.util.Date(captureTimeMillis)})"
              )
            }

            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
              val relativePath =
                if (isVideo) {
                  "Movies/Mentra"
                } else {
                  "Pictures/Mentra"
                }
              put(android.provider.MediaStore.MediaColumns.RELATIVE_PATH, relativePath)
              put(android.provider.MediaStore.MediaColumns.IS_PENDING, 1)
            }
          }

        val resolver = context.contentResolver
        val uri =
          resolver.insert(collection, values)
            ?: throw IllegalStateException("Failed to create MediaStore entry")

        try {
          resolver.openOutputStream(uri)?.use { outputStream ->
            file.inputStream().use { inputStream -> inputStream.copyTo(outputStream) }
          } ?: throw IllegalStateException("Failed to open output stream")

          if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
            values.clear()
            values.put(android.provider.MediaStore.MediaColumns.IS_PENDING, 0)
            resolver.update(uri, values, null, null)
          }

          android.util.Log.d(
            "CrustModule",
            "Successfully saved to gallery with proper DATE_TAKEN: ${file.name}"
          )
          mapOf("success" to true, "uri" to uri.toString())
        } catch (e: Exception) {
          resolver.delete(uri, null, null)
          throw e
        }
      } catch (e: Exception) {
        android.util.Log.e("CrustModule", "Error saving to gallery: ${e.message}", e)
        mapOf("success" to false, "error" to e.message)
      }
    }

    // MARK: - Navigation Commands (Google Navigation SDK)

    AsyncFunction("startNavigation") { lat: Double, lng: Double ->
      val activity = appContext.currentActivity
        ?: return@AsyncFunction mapOf("ok" to false, "error" to "no current activity (app backgrounded?)")

      val callbacks = object : NavigationManager.Callbacks {
        override fun onManeuver(payload: NavigationManager.ManeuverPayload) {
          sendEvent(
            "onNavManeuver",
            mapOf(
              "instruction" to payload.instruction,
              "roadName" to payload.roadName,
              "maneuverType" to payload.maneuverType,
              "distanceMeters" to payload.distanceMeters,
              "towardRoad" to payload.towardRoad,
              "nextManeuverType" to payload.nextManeuverType,
              "nextManeuverLabel" to payload.nextManeuverLabel,
            ),
          )
        }
        override fun onRerouting() {
          sendEvent("onNavRerouting", emptyMap<String, Any?>())
        }
        override fun onArrived() {
          sendEvent("onNavArrived", emptyMap<String, Any?>())
        }
        override fun onError(message: String) {
          sendEvent("onNavError", mapOf("message" to message))
        }
        override fun onLocation(payload: NavigationManager.LocationPayload) {
          sendEvent(
            "onNavLocation",
            mapOf(
              "lat" to payload.lat,
              "lng" to payload.lng,
              "accuracy" to payload.accuracy,
              "timestamp" to payload.timestamp,
            ),
          )
        }
        override fun onRoute(points: List<NavigationManager.RoutePoint>) {
          sendEvent(
            "onNavRoute",
            mapOf(
              "points" to points.map { mapOf("lat" to it.lat, "lng" to it.lng) },
            ),
          )
        }
      }

      activity.runOnUiThread {
        // 1) Verify ACCESS_FINE_LOCATION is granted to this process. The
        //    Google Nav SDK rejects getNavigator() with LOCATION_PERMISSION_MISSING
        //    even if location was granted in another module's context — it
        //    requires PackageManager.PERMISSION_GRANTED at call-time.
        val granted = androidx.core.content.ContextCompat.checkSelfPermission(
          activity,
          android.Manifest.permission.ACCESS_FINE_LOCATION,
        ) == android.content.pm.PackageManager.PERMISSION_GRANTED

        if (!granted) {
          // Request and abort this attempt. User taps Start again after granting.
          androidx.core.app.ActivityCompat.requestPermissions(
            activity,
            arrayOf(android.Manifest.permission.ACCESS_FINE_LOCATION),
            9001,
          )
          sendEvent(
            "onNavError",
            mapOf("message" to "ACCESS_FINE_LOCATION not granted — accept the prompt and tap Start again"),
          )
          return@runOnUiThread
        }

        // 2) Show T&C dialog (no-op after first acceptance), then start nav.
        com.google.android.libraries.navigation.NavigationApi
          .showTermsAndConditionsDialog(
            activity,
            "Mentra",
            object : com.google.android.libraries.navigation.NavigationApi.OnTermsResponseListener {
              override fun onTermsResponse(termsAccepted: Boolean) {
                if (!termsAccepted) {
                  sendEvent("onNavError", mapOf("message" to "user declined Google Nav T&C"))
                  return
                }
                NavigationManager.start(activity, lat, lng, callbacks)
              }
            },
          )
      }
      mapOf("ok" to true)
    }

    AsyncFunction("stopNavigation") {
      try {
        NavigationManager.stop()
        mapOf("ok" to true)
      } catch (e: Exception) {
        android.util.Log.e("CrustModule", "stopNavigation failed", e)
        mapOf("ok" to false, "error" to (e.message ?: "stop failed"))
      }
    }

    // MARK: - Heading (compass) — Android only

    AsyncFunction("startHeading") {
      val ctx = appContext.reactContext
        ?: appContext.currentActivity
        ?: return@AsyncFunction mapOf("ok" to false, "error" to "no context")
      HeadingManager.start(ctx, object : HeadingManager.Callback {
        override fun onHeading(degrees: Float) {
          sendEvent("onHeading", mapOf("degrees" to degrees.toDouble()))
        }
      })
      mapOf("ok" to true)
    }

    AsyncFunction("stopHeading") {
      HeadingManager.stop()
      mapOf("ok" to true)
    }
  }
}

File: mobile/modules/crust/ios/CrustModule.swift
import AVKit
import CoreLocation
import ExpoModulesCore
import GoogleNavigation
import Photos

/// User-visible album in Apple Photos for glasses sync (matches dedicated-folder behavior on Android).
private enum MentraSyncedMediaAlbum {
    static let localizedTitle = "Mentra"
}

public class CrustModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Crust")

        Constant("PI") {
            Double.pi
        }

        Events(
            "onChange",
            "phone_notification",
            "phone_notification_dismissed",
            "captions_tester_incident",
            "onNavManeuver",
            "onNavRerouting",
            "onNavArrived",
            "onNavError",
            "onNavOffRoute",
            "onNavLocation",
            "onNavRoute",
            "onHeading"
        )

        Function("hello") {
            "Hello world! 👋"
        }

        AsyncFunction("setValueAsync") { (value: String) in
            self.sendEvent("onChange", [
                "value": value,
            ])
        }

        AsyncFunction("requestNavigationPermission") { () -> [String: Any] in
            await withCheckedContinuation { continuation in
                NavigationManager.shared.requestPermission { accepted in
                    continuation.resume(returning: ["ok": true, "accepted": accepted])
                }
            }
        }

        AsyncFunction("startNavigation") { (lat: Double, lng: Double, options: [String: Any]?) -> [String: Any] in
            let simulate = options?["simulate"] as? Bool ?? false
            let speedMultiplier = options?["speedMultiplier"] as? Double ?? 1.0
            let mode = options?["mode"] as? String ?? "driving"

            var stops: [(lat: Double, lng: Double)] = []
            if let stopsArr = options?["stops"] as? [[String: Double]] {
                stops = stopsArr.compactMap { s in
                    guard let slat = s["lat"], let slng = s["lng"] else { return nil }
                    return (lat: slat, lng: slng)
                }
            }
            if stops.isEmpty { stops = [(lat: lat, lng: lng)] }

            return await withCheckedContinuation { continuation in
                NavigationManager.shared.start(
                    stops: stops,
                    mode: mode,
                    simulate: simulate,
                    speedMultiplier: speedMultiplier,
                    onEvent: { [weak self] payload in
                        guard let self else { return }
                        let kind = payload["kind"] as? String ?? ""
                        switch kind {
                        case "maneuver": self.sendEvent("onNavManeuver", payload)
                        case "rerouting": self.sendEvent("onNavRerouting", payload)
                        case "arrived": self.sendEvent("onNavArrived", payload)
                        case "off_route": self.sendEvent("onNavOffRoute", payload)
                        case "error": self.sendEvent("onNavError", payload)
                        default: break
                        }
                    },
                    onLocation: { [weak self] payload in
                        self?.sendEvent("onNavLocation", payload)
                    },
                    onRoute: { [weak self] payload in
                        self?.sendEvent("onNavRoute", payload)
                    }
                ) { ok, error in
                    var result: [String: Any] = ["ok": ok]
                    if let error { result["error"] = error }
                    continuation.resume(returning: result)
                }
            }
        }

        AsyncFunction("stopNavigation") { () -> [String: Any] in
            NavigationManager.shared.stop()
            return ["ok": true]
        }

        AsyncFunction("simulateDeviation") { (offsetMeters: Double?) -> [String: Any] in
            NavigationManager.shared.simulateDeviation(offsetMeters: offsetMeters ?? 50)
            return ["ok": true]
        }

        AsyncFunction("startHeading") { () -> [String: Any] in
            HeadingManager.shared.start { [weak self] degrees in
                self?.sendEvent("onHeading", ["degrees": degrees])
            }
            return ["ok": true]
        }

        AsyncFunction("stopHeading") { () -> [String: Any] in
            HeadingManager.shared.stop()
            return ["ok": true]
        }

        // Location:

        AsyncFunction("showLocationServicesDialog") { () -> Bool in
            return false
        }

        AsyncFunction("openLocationSettings") { () -> Bool in
            return false
        }

        AsyncFunction("openAppSettings") { () -> Bool in
            return false
        }

        AsyncFunction("openBluetoothSettings") { () -> Bool in
            return false
        }

        // MARK: - MentraOS Notification Commands

        AsyncFunction("setNotificationConfig") { (_: Bool, _: [String]) in
            // No-op on iOS
        }

        AsyncFunction("getInstalledApps") { () -> [[String: Any]] in
            return []
        }

        AsyncFunction("getInstalledAppsForNotifications") { () -> [[String: Any]] in
            return []
        }

        AsyncFunction("hasNotificationListenerPermission") { () -> Bool in
            return false
        }

        AsyncFunction("openNotificationListenerSettings") { () -> Bool in
            return false
        }

        // MARK: - Build Environment

        AsyncFunction("isBetaBuild") { () -> Bool in
            #if targetEnvironment(simulator)
                return false
            #else
                return Bundle.main.appStoreReceiptURL?.lastPathComponent == "sandboxReceipt"
            #endif
        }

        Function("showAVRoutePicker") { (tintColor: String?) in
            DispatchQueue.main.async {
                let picker = AVRoutePickerView()
                picker.prioritizesVideoDevices = false

                if let colorString = tintColor {
                    picker.tintColor = UIColor(hexString: colorString)
                } else {
                    picker.tintColor = .label
                }

                if let button = picker.subviews.first(where: { $0 is UIButton }) as? UIButton {
                    button.sendActions(for: .touchUpInside)
                }
            }
        }

        View(CrustView.self) {
            Prop("url") { (view: CrustView, url: URL) in
                if view.webView.url != url {
                    view.webView.load(URLRequest(url: url))
                }
            }

            Events("onLoad")
        }

        // MARK: - Image Processing Commands

        AsyncFunction("processGalleryImage") {
            (inputPath: String, outputPath: String, options: [String: Any]) -> [String: Any] in
            let lensCorrection = options["lensCorrection"] as? Bool ?? true
            let colorCorrection = options["colorCorrection"] as? Bool ?? true

            guard FileManager.default.fileExists(atPath: inputPath) else {
                return ["success": false, "error": "Input file does not exist"]
            }

            let processingTimeMs = ImageProcessor.process(
                inputPath: inputPath,
                outputPath: outputPath,
                lensCorrection: lensCorrection,
                colorCorrection: colorCorrection
            )

            if processingTimeMs >= 0 {
                return [
                    "success": true,
                    "outputPath": outputPath,
                    "processingTimeMs": processingTimeMs,
                ]
            } else {
                return ["success": false, "error": "Processing failed"]
            }
        }

        // MARK: - HDR Merge Commands

        AsyncFunction("mergeHdrBrackets") {
            (underPath: String, normalPath: String, overPath: String, outputPath: String)
            -> [String: Any] in
            let processingTimeMs = ImageProcessor.mergeHdr(
                underPath: underPath,
                normalPath: normalPath,
                overPath: overPath,
                outputPath: outputPath
            )
            if processingTimeMs >= 0 {
                return [
                    "success": true,
                    "outputPath": outputPath,
                    "processingTimeMs": processingTimeMs,
                ]
            } else {
                return ["success": false, "error": "HDR merge failed"]
            }
        }

        // MARK: - Video Stabilization Commands

        AsyncFunction("stabilizeVideo") {
            (inputPath: String, imuPath: String, outputPath: String) -> [String: Any] in
            guard FileManager.default.fileExists(atPath: inputPath) else {
                return ["success": false, "error": "Input video does not exist"]
            }
            guard FileManager.default.fileExists(atPath: imuPath) else {
                return ["success": false, "error": "IMU sidecar does not exist"]
            }

            let processingTimeMs = VideoStabilizer.stabilize(
                inputPath: inputPath,
                imuPath: imuPath,
                outputPath: outputPath
            )

            if processingTimeMs >= 0 {
                return [
                    "success": true,
                    "outputPath": outputPath,
                    "processingTimeMs": processingTimeMs,
                ]
            } else {
                return ["success": false, "error": "Stabilization failed"]
            }
        }

        // MARK: - Media Library Commands

        AsyncFunction("saveToGalleryWithDate") {
            (filePath: String, captureTimeMillis: Int64?) -> [String: Any] in
            let fileURL = URL(fileURLWithPath: filePath)

            guard FileManager.default.fileExists(atPath: filePath) else {
                return ["success": false, "error": "File does not exist"]
            }

            var assetIdentifier: String?
            let semaphore = DispatchSemaphore(value: 0)
            var resultError: Error?
            var creationFailed = false

            PHPhotoLibrary.shared().performChanges {
                let pathExtension = fileURL.pathExtension.lowercased()

                let creationRequest: PHAssetChangeRequest
                if ["mp4", "mov", "avi", "m4v"].contains(pathExtension) {
                    guard let req = PHAssetChangeRequest.creationRequestForAssetFromVideo(
                        atFileURL: fileURL
                    )
                    else {
                        NSLog("CrustModule: Failed to create video asset request for: \(filePath)")
                        creationFailed = true
                        return
                    }
                    creationRequest = req
                } else {
                    guard let req = PHAssetChangeRequest.creationRequestForAssetFromImage(
                        atFileURL: fileURL
                    )
                    else {
                        NSLog("CrustModule: Failed to create image asset request for: \(filePath)")
                        creationFailed = true
                        return
                    }
                    creationRequest = req
                }

                if let captureMillis = captureTimeMillis {
                    let captureDate = Date(
                        timeIntervalSince1970: TimeInterval(captureMillis) / 1000.0
                    )
                    creationRequest.creationDate = captureDate
                    NSLog("CrustModule: Setting creation date to: \(captureDate)")
                }

                guard let assetPlaceholder = creationRequest.placeholderForCreatedAsset else {
                    NSLog("CrustModule: Missing placeholder for created asset")
                    creationFailed = true
                    return
                }

                assetIdentifier = assetPlaceholder.localIdentifier

                let albumFetch = PHFetchOptions()
                albumFetch.predicate = NSPredicate(
                    format: "localizedTitle == %@", MentraSyncedMediaAlbum.localizedTitle
                )
                albumFetch.fetchLimit = 1

                let existingAlbums = PHAssetCollection.fetchAssetCollections(
                    with: .album,
                    subtype: .albumRegular,
                    options: albumFetch
                )

                if let album = existingAlbums.firstObject,
                   let albumChange = PHAssetCollectionChangeRequest(for: album)
                {
                    albumChange.addAssets([assetPlaceholder] as NSArray)
                } else if existingAlbums.firstObject == nil {
                    let newAlbumChange = PHAssetCollectionChangeRequest.creationRequestForAssetCollection(
                        withTitle: MentraSyncedMediaAlbum.localizedTitle
                    )
                    newAlbumChange.addAssets([assetPlaceholder] as NSArray)
                } else {
                    NSLog(
                        "CrustModule: Mentra album exists but is not writable; asset saved to library only"
                    )
                }
            } completionHandler: { _, error in
                resultError = error
                semaphore.signal()
            }

            semaphore.wait()

            if creationFailed {
                return ["success": false, "error": "Failed to create asset request - file may be corrupted or unsupported"]
            }

            if let error = resultError {
                NSLog("CrustModule: Error saving to gallery: \(error.localizedDescription)")
                return ["success": false, "error": error.localizedDescription]
            }

            NSLog("CrustModule: Successfully saved to gallery with proper creation date")
            return ["success": true, "identifier": assetIdentifier ?? ""]
        }

    }
}

extension UIColor {
    convenience init?(hexString: String) {
        var hex = hexString.trimmingCharacters(in: .whitespacesAndNewlines)
        hex = hex.replacingOccurrences(of: "#", with: "")

        var rgb: UInt64 = 0
        guard Scanner(string: hex).scanHexInt64(&rgb) else { return nil }

        let length = hex.count
        let r, g, b, a: CGFloat

        if length == 6 {
            r = CGFloat((rgb & 0xFF0000) >> 16) / 255.0
            g = CGFloat((rgb & 0x00FF00) >> 8) / 255.0
            b = CGFloat(rgb & 0x0000FF) / 255.0
            a = 1.0
        } else if length == 8 {
            r = CGFloat((rgb & 0xFF00_0000) >> 24) / 255.0
            g = CGFloat((rgb & 0x00FF_0000) >> 16) / 255.0
            b = CGFloat((rgb & 0x0000_FF00) >> 8) / 255.0
            a = CGFloat(rgb & 0x0000_00FF) / 255.0
        } else {
            return nil
        }

        self.init(red: r, green: g, blue: b, alpha: a)
    }
}
